### DISCUSSION TITLE
Using the fromModel method when the data collection is called

### DISCUSSION THREAD

**Discussion by matthiastjong** [spatie/laravel-data#382] [Q&A] [answered]
I have this data model:

```php
final class AdminCategoryData extends Data {
    public function __construct(
        #[DataCollectionOf(AdminComponentData::class)]
        public ?DataCollection $components,
    ) {}
}
```

Inside the `AdminComponentData`, I want to use the model `Component` itself to append some data:

```php
final class AdminComponentData extends Data {
    public function __construct(
        public string $location,
    ) {}

    public function fromModel (Component $component) : self {
        return new self(
            $component->getLocation(),
        );
    }
}
```

When I call the `AdminComponentData::fromModel($component)` I have access to the inner `Component` model and get the right location, however when it is called as a `#[DataCollectionOf(AdminComponentData::class)]`, that gets ignored and I get the exception that too little values are being sent in the data object..

FYI `Components` is a `hasMany` relationship on the `Category` model.

How do we make sure `fromModel` works when calling the model as a `DataCollection`?

The reason I am doing this because I want to append that value only on that specific `AdminComponentData` object. I will use the `Component` model in another Data Object where the `getLocation()` method is not needed, thus I do not want to append it to the whole `Component` model.

**Comment by matthiastjong** [comment, marked as answer]
For anyone that might run into the same challenge, this is how you fix it:

```php
final class AdminCategoryData extends Data
{
    public function __construct(
        #[DataCollectionOf(AdminComponentData::class)]
        public ?DataCollection $components,
    ) {
    }

    public function fromModel(RemoteCategory $category): self
    {
        return new self(
            AdminCategoryBrandData::collection($category->components->map(function ($component) {
                return AdminComponentData::fromModel($component);
            })),
        );
    }
}
```

You can send `fromModel` data objects to the collection, you just have to loop/map through them on the parent. 
If any package developer sees this I am wondering if this is the way to go?

**Comment by ricventu** [comment]
Hi,
you have to declare the method **static**

`public static function fromModel (Component $component) : self {`


and `AdminCategoryData::collect` will work using fromModel
